livre/lib/pipeline/RenderPipeline.cpp

```diff
@@ -82,40 +86,109 @@ struct RenderPipeline::Impl
+                             PipeFilter& sendHistogramFilter,
+                             Renderer& renderer,
+                             NodeAvailability& availability ) const
     {
-        PipeFilter visibleSetGenerator =
-                renderPipeline.add< VisibleSetGeneratorFilter >(
-                    "VisibleSetGenerator", _dataSource );
+        PipeFilterT< VisibleSetGeneratorFilter > visibleSetGenerator( "VisibleSetGenerator",
+                                                                      _dataSource );
 
-        createAndConnectUploaders( uploadPipeline,
-                                   visibleSetGenerator,
-                                   renderFilter );
+        visibleSetGenerator.getPromise( "Frustum" ).set( renderParams.frameInfo.frustum );
+        visibleSetGenerator.getPromise( "Frame" ).set( renderParams.frameInfo.timeStep );
+        visibleSetGenerator.getPromise( "DataRange" ).set( renderParams.renderDataRange );
+        visibleSetGenerator.getPromise( "Params" ).set( renderParams.vrParams );
+        visibleSetGenerator.getPromise( "Viewport" ).set( renderParams.pixelViewPort );
+        visibleSetGenerator.getPromise( "ClipPlanes" ).set( renderParams.clipPlanes );
 
-        for( size_t i = 0; i < nUploadThreads; ++i )
+        visibleSetGenerator.execute();
+
+        sendHistogramFilter.getPromise( "RelativeViewport" ).set( renderParams.viewport );
+        sendHistogramFilter.getPromise( "Id" ).set( renderParams.frameInfo.frameId );
+
+        const livre::UniqueFutureMap portFutures( visibleSetGenerator.getPostconditions( ));
+        const auto& nodeIds = renderer.order( portFutures.get< NodeIds >( "VisibleNodes" ),
+                                              renderParams.frameInfo.frustum );
+
+        const VolumeInformation& volInfo = _dataSource.getVolumeInfo();
+        const size_t blockMemSize = volInfo.maximumBlockSize.product() *
+                                    volInfo.getBytesPerVoxel() *
+                                    volInfo.compCount;
+
+        const uint32_t maxNodesPerPass =
+                renderParams.vrParams.getMaxGPUCacheMemoryMB() * LB_1MB / blockMemSize;
+
+        const uint32_t numberOfPasses = std::ceil( (float)nodeIds.size() / (float)maxNodesPerPass );
+
+        std::unique_ptr< boost::progress_display > showProgress;
+
+        if( numberOfPasses > 1 )
         {
-            std::stringstream name;
-            name << "DataUploader" << i;
-            PipeFilter uploader =
-                    static_cast< const livre::PipeFilter& >(
-                        uploadPipeline.getExecutable( name.str( )));
-            uploader.connect( "CacheObjects", histogramFilter, "CacheObjects" );
+            std::cout << "Multipass rendering. Number of passes: " << numberOfPasses <<std::endl;
+            showProgress.reset( new boost::progress_display( numberOfPasses ));
         }
+
+        for( uint32_t i = 0; i < numberOfPasses; ++i )
```

Comment:
this function is very long. Could the for-loop be refactored into its own function or something?